share/eds/eds.go

```diff
@@ -273,3 +277,59 @@ func ReadEDS(ctx context.Context, r io.Reader, root share.DataHash) (eds *rsmt2d
 	}
 	return eds, nil
 }
+
+// ConstructFn aliases a function that creates an ExtendedDataSquare.
+type ConstructFn func(
+	data [][]byte,
+	appVersion uint64,
+	squareSizeBounds int,
+	options ...nmt.Option,
+) (*rsmt2d.ExtendedDataSquare, error)
+
+// MakeExtendedDataSquare extends the given block data, returning the resulting
+// ExtendedDataSquare (EDS). If the data is empty,
+// nil is returned in place of the eds.
+func MakeExtendedDataSquare(
+	data [][]byte,
+	appVersion uint64,
+	squareSizeBounds int,
+	options ...nmt.Option,
+) (*rsmt2d.ExtendedDataSquare, error) {
+	if len(data) == 0 {
+		return nil, nil
+	}
+
+	// Construct the data square from the block's transactions
+	dataSquare, err := square.Construct(data, appVersion, squareSizeBounds)
+	if err != nil {
+		return nil, err
+	}
+	return extendShares(shares.ToBytes(dataSquare), options...)
+}
+
+func extendShares(s [][]byte, options ...nmt.Option) (*rsmt2d.ExtendedDataSquare, error) {
+	// Check that the length of the square is a power of 2.
+	if !shares.IsPowerOfTwo(len(s)) {
+		return nil, fmt.Errorf("number of shares is not a power of 2: got %d", len(s))
+	}
+	// here we construct a tree
+	// Note: uses the nmt wrapper to construct the tree.
+	squareSize := square.Size(len(s))
+	return rsmt2d.ComputeExtendedDataSquare(s,
+		appconsts.DefaultCodec(),
+		wrapper.NewConstructor(uint64(squareSize),
+			options...))
+}
+
+// StoreEDS will only store extended block if it is not empty and doesn't already exist.
+func StoreEDS(ctx context.Context, hash share.DataHash, eds *rsmt2d.ExtendedDataSquare, store *Store) error {
+	if eds == nil {
+		return nil
+	}
+	err := store.Put(ctx, hash, eds)
+	if errors.Is(err, dagstore.ErrShardExists) {
+		// block with given root already exists, return nil
+		return nil
+	}
+	return err
+}
```

Comment:
Why do you think this deserves a helper?

Reply:
You mean store Eds? It seems to me a really good function that stores  Eds if it does not exist in the store. We can re use in other places if needed. I did not want to insert this functionality inside edsMaker because I believe MakeEDS should take care of storing it - this is an absolutely another type of logic, imo

Reply:
Has check is also happening inside store.Put() call now. So you might want to replace this func with `store.Put()` everywhere.

Reply:
I believe another reason for this though was the eds nil check which was a bug at some point during the initial impl for bridge nodes, correct @renaynay ?